proof/invariant-abstract/AInvs.thy

```diff
@@ -1373,4 +1373,711 @@ lemma call_kernel_consumed_time_bounded:
+  apply (wpsimp wp: hoare_vcg_imp_lift')
+  done
+
+lemma cancel_all_ipc_schact_is_rct_imp_ct_activatable[wp]:
+  "cancel_all_ipc ep_ptr \<lbrace>\<lambda>s :: det_state. schact_is_rct s \<longrightarrow> ct_in_state activatable s\<rbrace>"
+  apply (clarsimp simp: cancel_all_ipc_def)
+  apply (wpsimp wp: update_sk_obj_ref_wp get_simple_ko_wp)
+  done
+
+lemma cancel_all_signals_schact_is_rct_imp_ct_activatable[wp]:
+  "cancel_all_signals ntfnptr \<lbrace>\<lambda>s :: det_state. schact_is_rct s \<longrightarrow> ct_in_state activatable s\<rbrace>"
+  apply (clarsimp simp: cancel_all_signals_def)
+  apply (wpsimp wp: get_simple_ko_wp)
+  done
+
+lemma set_ntfn_obj_ref_schact_is_rct_imp_ct_activatable[wp]:
+  "set_ntfn_obj_ref update ref new \<lbrace>\<lambda>s :: det_state. schact_is_rct s \<longrightarrow> ct_in_state activatable s\<rbrace>"
+  apply (wpsimp wp: update_sk_obj_ref_wp)
+  apply (clarsimp simp: ct_in_state_def pred_tcb_at_def obj_at_def simple_obj_at_def)
+  done
+
+lemma unbind_maybe_notification_schact_is_rct_imp_ct_activatable[wp]:
+  "unbind_maybe_notification ntfnptr \<lbrace>\<lambda>s :: det_state. schact_is_rct s \<longrightarrow> ct_in_state activatable s\<rbrace>"
+  apply (clarsimp simp: unbind_maybe_notification_def)
+  apply (wpsimp wp: update_sk_obj_ref_wp set_tcb_obj_ref_wp get_sk_obj_ref_wp)
+  apply (fastforce simp: ct_in_state_def pred_tcb_at_def obj_at_def)
+  done
+
+lemma sched_context_maybe_unbind_ntfn_schact_is_rct_imp_ct_activatable[wp]:
+  "sched_context_maybe_unbind_ntfn ntfn_ptr
+   \<lbrace>\<lambda>s :: det_state. schact_is_rct s \<longrightarrow> ct_in_state activatable s\<rbrace>"
+  apply (clarsimp simp: sched_context_maybe_unbind_ntfn_def)
+  apply (wpsimp wp: update_sk_obj_ref_wp get_sk_obj_ref_wp update_sched_context_wp)
+  apply (fastforce simp: ct_in_state_def pred_tcb_at_def obj_at_def)
+  done
+
+crunches fast_finalise
+  for schact_is_rct_imp_ct_activatable[wp]:  "\<lambda>s :: det_state. schact_is_rct s \<longrightarrow> ct_in_state activatable s"
+
+lemma cap_delete_one_schact_is_rct_imp_ct_activatable[wp]:
+  "cap_delete_one slot \<lbrace>\<lambda>s :: det_state. schact_is_rct s \<longrightarrow> ct_in_state activatable s\<rbrace>"
+  apply (clarsimp simp: cap_delete_one_def unless_def)
+  apply (rule hoare_seq_ext_skip, wpsimp)
+  apply (rule hoare_when_cases, simp)
+  apply (rule hoare_seq_ext_skip, solves wpsimp)+
+  apply (wpsimp wp: hoare_vcg_imp_lift')
+  done
+
+crunches deleting_irq_handler
+  for schact_is_rct_imp_ct_activatable[wp]: "\<lambda>s :: det_state. schact_is_rct s \<longrightarrow> ct_in_state activatable s"
+
+lemma finalise_cap_schact_is_rct_imp_ct_activatable:
+  "finalise_cap cap final
+   \<lbrace>\<lambda>s :: det_state . (schact_is_rct s \<longrightarrow> ct_in_state activatable s)
+                      \<and> (\<exists>slot. cte_wp_at ((=) cap) slot s) \<and> invs s\<rbrace>"
```

Comment:
This postcondition seems a bit unusual. Out of curiosity, do you remember why you needed to prove them together like this?

Reply:
I was able to rephrase the preconditions to be a little bit more usual. The changes are in the last commit